From a32b101bc1b77df2cd7830dda0b309886c277b62 Mon Sep 17 00:00:00 2001
From: Andre Klapper <a9016009@gmx.de>
Date: Fri, 10 Jan 2025 11:51:08 +0100
Subject: [PATCH 11/14] JSON rendering: Avoid exception iterating on
 non-iterable objects

Summary:
Check for `is_iterable($object)` (available since PHP 7.1) to avoid an exception calling `foreach` on `$object` in the rare case that the object is a custom Phorge class (`PHUIBoxView`, `PhutilSafeHTML`) and not an array. cf https://we.phorge.it/rARC7570dd0da119627ff83bc6db3be06b51eb5b366b for a similar patch to handle PHP stdClass objects.

See downstream https://phabricator.wikimedia.org/T373316.

Test Plan:
* Unclear to reproduce the actual issue, likely something Feed related.
* Open an existing JSON Paste at http://phorge.localhost/P1 and see that it still renders.
* Open an existing JSON File at http://phorge.localhost/F1 and see that it still renders.

Reviewers: O1 Blessed Committers, valerio.bozzolan

Reviewed By: O1 Blessed Committers, valerio.bozzolan

Subscribers: tobiaswiese, valerio.bozzolan, Matthew, Cigaryno

Differential Revision: https://we.phorge.it/D25855
---
 src/parser/PhutilJSON.php | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/parser/PhutilJSON.php b/src/parser/PhutilJSON.php
index dd6c2fb638c7..75bf616abcd2 100644
--- a/src/parser/PhutilJSON.php
+++ b/src/parser/PhutilJSON.php
@@ -51,7 +51,7 @@ final class PhutilJSON extends Phobject {
       $object = (array)$object;
     }
 
-    if (empty($object)) {
+    if (empty($object) || !is_iterable($object)) {
       return '{}';
     }
 
-- 
2.48.1

